urwid/_async_kw_event_loop.py

```diff
@@ -0,0 +1,244 @@
+        """
+        return self._cancel_scope(handle)
+
+    def _cancel_scope(self, scope):
+        """Cancels the given Trio cancellation scope.
+
+        Returns:
+            True if the scope was cancelled, False if it was cancelled already
+            before invoking this function
+        """
+        scope, cancelled = scope
+        existed = not cancelled[0]
+        cancelled[0] = True
+        scope.cancel()
+        return existed
+
+    def _create_cancel_scope(self):
+        """Creates a Trio cancellation scope and a corresponding mutable flag
+        that indicates whether the scope was cancelled already _from_ urwid.
+
+        This is needed because `CancelScope.cancelled_caught` stays `False` until
+        someone actually _handles_ the cancellation, and
+        `CancelScope.cancel_called` can only be called from an async context
+        (which we cannot guarantee).
+        """
+        return self._trio.CancelScope(), [False]
+
+    def run(self):
+        """Starts the event loop. Exits the loop when any callback raises an
+        exception. If ExitMainLoop is raised, exits cleanly.
+        """
+        def _exception_handler(exc):
+            if isinstance(exc, ExitMainLoop):
+                return None
+            else:
+                return exc
+
+        with self._trio.MultiError.catch(_exception_handler):
+            if not self._nursery:
+                self._trio.run(self._main_task)
+            else:
+                self._nursery.start_soon(self._main_task)
+
+    def watch_file(self, fd, callback):
+        """Calls `callback()` when the given file descriptor has some data
+        to read. No parameters are passed to the callback.
+
+        Parameters:
+            fd: file descriptor to watch for input
+            callback: function to call when some input is available
+
+        Returns:
+            a handle that may be passed to `remove_watch_file()`
+        """
+        return self._start_task(self._watch_task, fd, callback)
+
+    async def _alarm_task(self, scope, seconds, callback):
+        """Asynchronous task that sleeps for a given number of seconds and then
+        calls the given callback.
+
+        Parameters:
+            scope: the cancellation scope that can be used to cancel the task
+            seconds: the number of seconds to wait
+            callback: the callback to call
+        """
+        with scope:
+            await self._sleep(seconds)
+            callback()
+
+    async def _idle_task(self):
+        """Asynchronous task that sleeps for a short amount of time and then
+        calls all the registered idle callbacks.
+
+        Used to simulate idle callbacks in the Trio event loop that has no
+        concept of being idle.
+        """
+        while True:
+            await self._sleep(self._idle_emulation_delay)
+            for idle_callback in self._idle_callbacks.values():
+                idle_callback()
```

Comment:
It's a bit unfortunate that this needs to wake up constantly, even if there isn't actually any work to do; it hurts battery life b/c it stops CPUs from going to sleep. I think it's fine for a first version, but I do wonder if there's any way to avoid that.

At first I thought well, maybe we can only run the loop while there's an idle function actually registered. But I checked and it looks like urwid always has an idle function registered, so that wouldn't actually help anything.

The reason urwid always has an idle function registered is that it uses it to redraw the screen. Which struck me as a weird thing to do from an idle callback, since normally you'd only want to redraw the screen after something changed, not when the program is sitting there idle!

After looking a bit more I think what's going on here is:

urwid's own home-grown loop calls the idle callbacks *exactly once before each time it sleeps for an extended period*. So the idea is that you go back and forth between executing code, and sleeping; and we assume that the screen might change while executing code, but not while sleeping, so we want to update the screen on those transitions. It's a little bit hacky compared to a more fine-grained system that actually tracks whether the screen needs redraw (e.g. it might do extra redraws, or if the program is busy for a while then it might not redraw for a while), but it's a lot simpler to implement than fine-grained damage tracking, and probably works about 99% as well in practice, so it makes sense.

However, other main loops don't actually provide this operation. E.g. Glib has something called an "idle callback", but it runs constantly whenever the loop could otherwise be sleeping, which is very different. But, since they're both named "idle callback", maybe no-one noticed?

So like... maybe the real solution here is for urwid to implement a fine-grained damage tracking system, which might be non-trivial, but I think it's the only way to really properly support external event loops.

In the mean time...... there *is* a way to get the original run-once-before-sleeping semantics in Trio. It doesn't fix the issue for other loops like glib, and it's pretty gross, but maybe it's a reasonable medium-term hack.

What you could do is use [Trio's instrumentation API](https://trio.readthedocs.io/en/latest/reference-hazmat.html#instrument-api). Normally this is intended for debugging and profiling, but it does let you find out when Trio is [about to go to sleep and run some code](https://trio.readthedocs.io/en/latest/reference-hazmat.html#trio.abc.Instrument.before_io_wait).

Because they're a low-level debugging interface, instrumentation hooks run in a weird context: you have to be very careful not to let any exceptions escape, and you're not inside any task so you can't `await` or use the cancellation machinery normally. (Though you should still be able to do operations that rely on run-global state, like call `trio.current_time()`.)

Anyway, like I said, I don't think you necessarily need to do anything about this now. But it's interesting to think about the options...

Reply:
> It's a bit unfortunate that this needs to wake up constantly, even if there isn't actually any work to do

Yes, I felt bad about this for a while until I realized that `urwid` registers an idle callback for redrawing the screen anyway so there's always at least one idle callback. Other async event loops like the one for Twisted or `asyncio` also handle things in a similar way and I did not know that Trio's event loop actually has a concept similarly to the idle callback of `urwid`'s default `select`-based loop.

I'll investigate whether I could use the instrumentation API for the event loop. In the previous PR, you suggested using `wait_all_tasks_blocked()` instead - wouldn't that be a better option? (At least in the sense that we don't need to register a custom instrument).

Reply:
> In the previous PR, you suggested using wait_all_tasks_blocked() instead - wouldn't that be a better option? (At least in the sense that we don't need to register a custom instrument).

Not really, unfortunately – if you wrote a loop like:

```python
while True:
    await wait_all_tasks_blocked()
    ...
```

...then the loop would indeed wait until the program was otherwise idle before starting to run, *but* once the rest of the program became idle then it would keep running constantly, similarly to the loop you have now.

Reply:
Okay, I have tried to use the instrumentation API (haven't committed it yet), and it seems to work - at least the unit tests pass correctly. Would `before_io_wait()` also get called if there are no IO tasks to wait for at all and there are only pending `sleep()` calls that might wake up the Trio loop?

Reply:
> Would before_io_wait() also get called if there are no IO tasks to wait for at all and there are only pending sleep() calls that might wake up the Trio loop?

Yeah... the name is maybe a bit confusing, actually. Operating systems provide a single combined check-for-I/O-and-maybe-wait-for-timeouts-to-expire operation, so Trio always uses that. Sometimes Trio just wants to check for I/O without waiting (because there are other tasks already ready to run), sometimes it just wants to wait for timeouts, sometimes it wants to do both at once – in all cases it uses the same underlying operation. And in all cases it calls  `before_io_wait`.

You can tell whether it's going to sleep, actually. If the `timeout` passed in to `before_io_wait` is zero, then it means there are already tasks ready to run, and Trio is just doing a quick check to see if any I/O has appeared – it isn't actually going to wait, despite the name. If the `timeout` argument is > zero, then it means Trio is ready to sleep for a bit.

I guess to really match the traditional urwid semantics, you could make your idle callbacks only run when `timeout > 0`. I'm not sure if this is actually better or not – the tradeoff is that we do fewer screen refreshes, which is more efficient (Trio often does a quick check for I/O even though the program is otherwise busy), but if the program is busy for a while then it could cause the screen refreshes to get out of date.

Reply:
Okay, I have updated my PR with your suggestions - thanks a lot!